src/security/stats.cpp

```diff
@@ -0,0 +1,227 @@
+#include <string>
+#include <vector>
+
+extern "C" {
+#include <netdb.h>
+#include <ngx_cycle.h>
+#include <sys/socket.h>
+#include <unistd.h>
+}
+
+using namespace std::literals::string_view_literals;
+
+namespace {
+
+int create_udp_socket(const char *hostname, const char *port,
+                      struct sockaddr_storage &server_address) {
+  struct addrinfo hints = {
+      .ai_flags = AI_ALL,
+      .ai_family = AF_UNSPEC,
+      .ai_socktype = SOCK_DGRAM,
+  };
+
+  struct addrinfo *result;
+  if (getaddrinfo(hostname, port, &hints, &result) != 0) {
+    return -1;
+  }
+
+  int sockfd = -1;
+  for (struct addrinfo *rp = result; rp != nullptr; rp = rp->ai_next) {
+    sockfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+    if (sockfd == -1) {
+      continue;
+    }
+
+    memcpy(&server_address, rp->ai_addr, rp->ai_addrlen);
+  }
+
+  freeaddrinfo(result);
+
+  return sockfd;
+}
+
+#ifdef __GLIBC__
+#include <malloc.h>
+#else
+extern "C" struct mallinfo2 {  // NOLINT
+  size_t arena;                /* Non-mmapped space allocated (bytes) */
+  size_t ordblks;              /* Number of free chunks */
+  size_t smblks;               /* Number of free fastbin blocks */
+  size_t hblks;                /* Number of mmapped regions */
+  size_t hblkhd;               /* Space allocated in mmapped regions (bytes) */
+  size_t usmblks;              /* See below */
+  size_t fsmblks;              /* Space in freed fastbin blocks (bytes) */
+  size_t uordblks;             /* Total allocated space (bytes) */
+  size_t fordblks;             /* Total free space (bytes) */
+  size_t keepcost;             /* Top-most, releasable space (bytes) */
+};
+#endif  // __GLIBC__
+
+template <typename Send>
+void report_memory_stats(std::string_view pid, Send &&send) {
+  static bool checked = false;
+  static struct mallinfo2 (*mallinfo2_fn)(void) = nullptr;
+
+  if (!checked) {
+    mallinfo2_fn = (struct mallinfo2(*)(void))dlsym(RTLD_DEFAULT, "mallinfo2");
+    checked = true;
+  }
+
+  if (!mallinfo2_fn) {
+    return;  // mallinfo2 not available
+  }
+
+  struct mallinfo2 info = mallinfo2_fn();
+
+  std::invoke(std::forward<Send>(send), "memory.arena"sv, 'g', info.arena);
+  std::invoke(std::forward<Send>(send), "memory.uordblks"sv, 'g',
+              info.uordblks);
+  std::invoke(std::forward<Send>(send), "memory.fordblks"sv, 'g',
+              info.fordblks);
+  std::invoke(std::forward<Send>(send), "memory.hblkhd"sv, 'g', info.hblkhd);
+}
+
+std::string_view prepare_metric(const std::string_view metric,
+                                std::string_view pid, char type,
+                                std::uint32_t val) {
+  static constexpr size_t max_payload_len =
+      sizeof("appsec.tasks_submission_failed_65535:18446744073709551615|c") - 1;
+
+  static std::array<char, max_payload_len> buf{};
+
+  char *it = std::copy(metric.begin(), metric.end(), buf.begin());
+  *it = '_';
+  it++;
+  it = std::copy(pid.begin(), pid.end(), it);
```

Comment:
nit: Any particular reason to put the pid in the metric name rather than a metric tag?